microcloud/service/lxd.go

```diff
@@ -148,7 +148,21 @@ func (s LXDService) Bootstrap() error {
 		return fmt.Errorf("Failed to initialize cluster: %w", err)
 	}
 
-	return nil
+	for {
+		select {
+		case <-time.After(30 * time.Second):
+			return fmt.Errorf("Timed out waiting for LXD cluster to initialize")
+		default:
+			names, err := client.GetClusterMemberNames()
```

Comment:
Is this a busy wait that will dos the server until its ready? Shall we put a short skeeo between esxh iteration?

Reply:
Actually I don't see how this loop can work as the default case will always immediately be chosen and then block on GetClusterMemberNames. I don't see how the timeout can work. 

Did I miss something?

Reply:
Ideally we need GetClusterMemberNames to take a context, is this in the lxd client ?

Reply:
This isn't trying to block fetching cluster members. It's repeatedly fetching the cluster members over the course of 30 seconds until the list is no longer empty. 

The reason for this is that there seems to be an bug (which I will file an issue for once I've investigated it a bit more) where LXD in particular will occasionally return from bootstrapping while the cluster is still non-functional.

Reply:
I ended up adding the loop to all the services just to keep things consistent

Reply:
It could use short sleep, though.

Reply:
I don't think adding a context timeout to `ClusterMemberNames` all that necessary in this case because the client should already timeout before 30s is over.

Reply:
If you've set up MicroCloud and received the error `failed to join the cluster with the given token` or something like that, then it's because of this error. When issuing a token after bootstrapping, the list of cluster members is still empty so the token gets generated without any addresses.

Reply:
Ah yes I see, makes sense now. I thought you were trying to timeout the call to `ClusterMemberNames` - this is why I shouldnt do reviews so late :)

Anyway, you're right we need a sleep in there otherwise we are going to be sending a DoS to the target server.

But still something isn't quite right with this loop. The time.After call is repeated on each loop iteration, meaning a new time.After channel and go routine is created for each iteration. This is a problem for 2 reason; 1 its resetting the timer on each check, and 2 its creating a new timer every time.

From the Go manual:

> The underlying Timer is not recovered by the garbage collector until the timer fires. If efficiency is a concern, use NewTimer instead and call Timer.Stop if the timer is no longer needed. 

https://pkg.go.dev/time#After

So I think it would be more correct and efficient to do this:

```go
timeout := time.NewTimer(30 * time.Second)
defer timeout.Stop()

for {
		select {
		case <-timeout.C:
			return fmt.Errorf("Timed out waiting for LXD cluster to initialize")
		default:
			names, err := client.GetClusterMemberNames()
           ....
          time.Sleep(time.Second)
}
```

An alternative is to use a context here:

```go
ctx, cancel := context.WithTimeout(context.TODO, 30*time.Second)
defer cancel()

for {
		select {
		case <-ctx.Done():
			return fmt.Errorf("Timed out waiting for LXD cluster to initialize: %w", ctx.Error())
		default:
			names, err := client.GetClusterMemberNames()
           ....
         time.Sleep(time.Second)
}
```

What do you think?